internal/compiler/passes/remove_return.rs

```diff
@@ -0,0 +1,378 @@
+            ExpressionResult::MaybeReturn {
+                mut pre_statements,
+                condition,
+                returned_value,
+                actual_value,
+            } => {
+                pre_statements.push(Expression::Condition {
+                    condition: condition.into(),
+                    true_expr: actual_value.unwrap_or(Expression::CodeBlock(vec![])).into(),
+                    false_expr: returned_value.unwrap_or(Expression::CodeBlock(vec![])).into(),
+                });
+                Expression::CodeBlock(pre_statements)
+            }
+            ExpressionResult::ReturnObject { value, has_value, has_return_value } => {
+                static COUNT: std::sync::atomic::AtomicUsize =
+                    std::sync::atomic::AtomicUsize::new(0);
+                let name = format!(
+                    "returned_expression{}",
+                    COUNT.fetch_add(1, std::sync::atomic::Ordering::Relaxed)
+                );
+                let load =
+                    Box::new(Expression::ReadLocalVariable { name: name.clone(), ty: value.ty() });
+                Expression::CodeBlock(vec![
+                    Expression::StoreLocalVariable { name: name.into(), value: value.into() },
+                    Expression::Condition {
+                        condition: Expression::StructFieldAccess {
+                            base: load.clone(),
+                            name: FIELD_CONDITION.into(),
+                        }
+                        .into(),
+                        true_expr: if has_value {
+                            Expression::StructFieldAccess {
+                                base: load.clone(),
+                                name: FIELD_ACTUAL.into(),
+                            }
+                        } else {
+                            Expression::default_value_for_type(ty)
+                        }
+                        .into(),
+                        false_expr: if has_return_value {
+                            Expression::StructFieldAccess {
+                                base: load.clone(),
+                                name: FIELD_RETURNED.into(),
+                            }
+                        } else {
+                            Expression::default_value_for_type(ty)
+                        }
+                        .into(),
+                    },
+                ])
+            }
+        }
+    }
+
+    fn into_return_object(self, ty: &Type, ret_ty: &Type) -> Expression {
+        match self {
+            ExpressionResult::Just(e) => make_struct(
+                [
+                    (FIELD_CONDITION, Type::Bool, Expression::BoolLiteral(true)),
+                    (
+                        FIELD_RETURNED.into(),
+                        ret_ty.clone(),
+                        Expression::default_value_for_type(ret_ty),
+                    ),
+                    (FIELD_ACTUAL, e.ty(), e),
```

Comment:
I understand that implementation wise the order doesn't matter here, but for readability I suggest to sync it with the order of the ReturnObject, where the actual field comes before the returned value field.

```suggestion
                    (FIELD_ACTUAL, e.ty(), e),
                    (
                        FIELD_RETURNED.into(),
                        ret_ty.clone(),
                        Expression::default_value_for_type(ret_ty),
                    ),
```